### ISSUE TITLE
Use FileRef constructor passing IOStream

### ISSUE THREAD

**Issue by fillobotto** [taglib/taglib#796]
I'm using TagLib in an Android project, building it in arm mode and it's all working.

I'm trying to manipulate audios without passing file paths, but passing to the native code a the `byte[]` which is then converted to `char *`.
I double checked the integrity of the bytes while passing them to the native code, but when I initialize the `FileRef`, it always fails again the `isNull()` check.

```
char* buf = new char[ numBytes ];
env->GetByteArrayRegion(data, 0, numBytes, reinterpret_cast<jbyte*>(buf));

ByteVector byteVector = ByteVector::fromCString(buf, numBytes);
TagLib::IOStream *vectorStream = new TagLib::ByteVectorStream(byteVector);
TagLib::FileRef f(vectorStream);

if(f.isNull())
     return;
```

**Comment by TsudaKageyu** [comment]
It's a flaw caused by a past change that enables `FileRef` to take a stream. 
`FileRef` requires a file extension to determine the file type, but `ByteVectorStream` can't hold a file name.

A possible solution is to enable `ByteVectorStream` to take a file name via its constructor.

**Comment by fillobotto** [comment]
I would appreciate if someone could archieve this task with a pull request. Unfortunately I'm not good at c++ and I'm using it only as part of an Android Java project.

**Comment by scotchi** [comment]
This seems very wrong.  The file name usage in FileRef was always considered only a default (i.e. non-production) implementation of mime type detection.  Pushing that (bad) implementation out to streams, which have no inherent file name seems very incorrect.  For this to work properly (and this should probably just be updated in the docs), the FileTypeResolver would need to be fixed so that it properly detects the mime type from the stream.  If a helper is going to be added, it should be something with a mime type, not with a file name.

**Comment by TsudaKageyu** [comment]
Sorry, I missed out on the comment in `fileref.h`. I will close the PR.
I'm going to think more how to implement the stream based file type detection.

**Comment by fillobotto** [comment]
Thank you for the effort, I will wait for a solution because I should have to write a temporary file instead, which is very resource-consuming on mobiles.

**Comment by TsudaKageyu** [comment]
I'm thinking about how to detect the file type of a stream not depending on its file name. 
Now I think that it's not so hard to write a stream based file type resolver except for some formats that don't have keys to identify them. Specifically, Old MusePak, MOD, IT, S3M, XM.

AIFF: Starts with "FORM????AIFF" or "FORM????AIFC"
APE: Has "MAC ". (ID3v2 tag may precede.)
ASF: Starts with designated GUID.
FLAC: Has "fLaC". (ID3v2 tag may precede.)
MP4: Has "ftyp" at the offset 4. 
MusePak: Newer files start with "MPCK" or "MP+".
MPEG: Has an MPEG frame. (ID3v2 tag may precede.)
Ogg Vorbis: Has "OggS". First page starts with "\x01vorbis"
Ogg FLAC: Has "OggS". First page starts with "fLaC"
Ogg Opus: Has "OggS". First page starts with "OpusHead"
Ogg Opus: Has "OggS". First page starts with "Speex   "
WAV: Starts with "RIFF????WAVE"
TrueAudio: Starts with "TTA". (ID3v2 tag may precede.)
WavPack: Starts with "wvpk".

**Comment by scotchi** [comment]
Yeah, and most MIME systems have a resolver system where every `File` subclass would register itself -- so, for example, `File::MPEG` would have a static boolean method that it would register with a file type resolver.  That makes it both overridable and extensible.

**Comment by fillobotto** [comment]
Would it also be possible to open a `TagLib::IOStream` from a `FILE*` object?

**Comment by ufleisch** [comment]
Fixed by #799.
